library/proc_macro/src/bridge/mod.rs

```diff
@@ -449,3 +464,16 @@ compound_traits!(
         Literal(tt),
     }
 );
+
+/// Context provided alongside the initial inputs for a macro expansion.
+/// Provides values such as spans which are used frequently to avoid RPC.
+#[derive(Clone)]
+struct ExpnContext<S> {
+    def_site: S,
+    call_site: S,
+    mixed_site: S,
+}
```

Comment:
I've been thinking about this a bunch and something like this might be good:
```rust
struct ExpandRequest<Span, I> {
    def_site: Span,
    call_site: Span,
    mixed_site: Span,
    input: I,
}
```
and then `ExpandRequest<S::Span, S::TokenStream>` and `ExpandRequest<S::Span, (S::TokenStream, S::TokenStream)>` would be used for the two kinds of proc macros.

And if that type is actually used in the API, the `Context` trait wouldn't be needed.

I'm not super happy with the names, but this kind of "request" type would fit with some other ideas around interacting with a proc macro crate as a whole with S->C "requests" (at that point the client/server terminology sadly starts getting in the way, but I hope it's clear what I'm referring to).

Reply:
I have been planning to use the `Context` trait more in part 4 (#98189). It's being used for handling the serialization/deserialization of `Symbol` types, as well as to provide a free function for validating `Ident`s without requiring the `proc-macro` crate to depend on `unicode_xid` and `unicode-normalization`. I could introduce a similar trait in part 4 for that purpose though, and move the spans off of it in this part.

One downside of moving this directly into the API is that it will move more things out of the shared proc macro server and into the 3 seperate proc macro wrapper implementations ([1](https://github.com/rust-lang/rust/blob/1aabd8a4a6e1871f14e804302bd60dfcbffd5761/compiler/rustc_expand/src/proc_macro.rs#L34-L40), [2](https://github.com/rust-lang/rust/blob/1aabd8a4a6e1871f14e804302bd60dfcbffd5761/compiler/rustc_expand/src/proc_macro.rs#L63-L71), [3](https://github.com/rust-lang/rust/blob/1aabd8a4a6e1871f14e804302bd60dfcbffd5761/compiler/rustc_expand/src/proc_macro.rs#L109-L119)), as we'd need to do the dance to get the spans there, or continue to compute them [within the server type](https://github.com/rust-lang/rust/blob/1aabd8a4a6e1871f14e804302bd60dfcbffd5761/compiler/rustc_expand/src/proc_macro_server.rs#L357-L361), then pull them out to pass alongside, though we can probably get away with that.

Reply:
We discussed this a bit async, and agreed to move the methods to the `Server` trait instead of the `Context` trait. We'll revisit changing the shape of the `Client` API to make context passing as a parameter more explicit in the future, but it'll be a bit tricky to do in this patch.